### ISSUE TITLE
[all] support contexts

### ISSUE THREAD

**Issue by jayshrivastava** [apache/pulsar-client-go#1170]
**Is your feature request related to a problem? Please describe.**
The library does not take a context/context. 

**Describe the solution you'd like**
I expect clients and producers to take a context argument upon creation and have all goroutines created by the clients and producers take a context as well. This is important for tracing (see https://medium.com/swlh/distributed-tracing-for-go-microservice-with-opentracing-1fc1aec76b3e). The context needs to be piped into all goroutines for observability. 

It's also important to select on `ctx.Done()` when sending or recieving to channels. The libary does not do that at the moment. Ex.
What if the command chan is full and `cp` cannot be send? This method will block. It would be beneficial to also select on a context provided by the user of the library in case the caller wants to cancel long running blocked operations. 

https://github.com/apache/pulsar-client-go/blob/f476814a9e1bac484e9589d4fb3f299066f54ddc/pulsar/producer_partition.go#L1462-L1473

**Comment by merlimat** [comment]
@jayshrivastava Yes, you're right, it would be really good to take those contexts (and use them). 

The main difficulty, as in #1165, is how to do that without breaking compatibility. 

Or maybe, it would be a good opportunity to break compatibility, and change versioning scheme to 1.0. Would you like to drive a discussion on the dev@pulsar.apache.org mailing list about this?
